core/tests/protractor_utils/ExplorationEditorPage.js

```diff
@@ -215,6 +215,11 @@ var ExplorationEditorPage = function() {
 
   this.saveChanges = async function(commitMessage) {
     var toastSuccessElement = element(by.css('.toast-success'));
+    // This is necessary to give the page time to record the changes,
+    // so that it does not attempt to stop the user leaving.
+    await waitFor.invisibilityOf(
+      toastSuccessElement,
```

Comment:
Why are we waiting for the invisibility of the toast before clicking the "Save button"?

I think earlier the structure was better i.e, complete the save saveChanges call after the toast is disappeared. Any specific reason for moving it to the top?

Reply:
I think, I wrote this code to do a "visbility of" then an "invisibility of" to avoid the case where there is a toast from a previous save and the test mistakes a stale one to be the current one. Moving the invisibility of to the top preserves this behaviour.

Another reason I wanted to do this is to avoid waiting for the toast to disappear because it will just slow down the test (now that we are waiting for a longer duration).

Reply:
> I think, I wrote this code to do a "visbility of" then an "invisibility of" to avoid the case where there is a toast from a previous save and the test mistakes a stale one to be the current one. Moving the invisibility of to the top preserves this behaviour.

But as per our earlier structure, we would never be in a situation where we have toast from the previous save, right?

> Another reason I wanted to do this is to avoid waiting for the toast to disappear because it will just slow down the test (now that we are waiting for longer duration).

I think the save function should complete only after the toast disappears will help us complete one action and reaction at a time. Note, other functions/actions can also raise a toast and they are not waiting for the invisibility of the old toast (In general all such functions/actions should wait for the toast to disappear.).

(Sorry I have less context of the issue here so ignore the next comment if it's irrelevant) Also, in general, I think we should focus on fixing the e2e test issues instead of changing the user-facing behavior to fix the e2e flakiness. Let me know what you think? :)

Reply:
> But as per our earlier structure, we would never be in a situation where we have toast from the previous save, right?

This one as well but it is specific to toast messages coming from continuous saves. But I understand your concern that other functions also wait for invisibility of saves. I will move the check back, thanks!

@DubeySandeep, yes, I agree it is not great to change the main code to fix an e2e test. Do you have any suggestion on how to fix this without doing this? I'm happy to discuss a better solution.

Reply:
Created https://github.com/oppia/oppia/pull/11846.